haystack_experimental/components/generators/chat/openai.py

```diff
@@ -0,0 +1,426 @@
+            chunks: List[StreamingChunk] = []
+            chunk = None
+
+            # pylint: disable=not-an-iterable
+            for chunk in chat_completion:
+                if chunk.choices and streaming_callback:
+                    chunk_delta: StreamingChunk = self._build_chunk(chunk)
+                    chunks.append(chunk_delta)
+                    streaming_callback(chunk_delta)  # invoke callback with the chunk_delta
+            completions = [self._connect_chunks(chunk, chunks)]
+        # if streaming is disabled, the completion is a ChatCompletion
+        elif isinstance(chat_completion, ChatCompletion):
+            completions = [self._build_message(chat_completion, choice) for choice in chat_completion.choices]
+
+        # before returning, do post-processing of the completions
+        for message in completions:
+            self._check_finish_reason(message)
+
+        return {"replies": completions}
+
+    def _connect_chunks(self, chunk: Any, chunks: List[StreamingChunk]) -> ChatMessage:
+        """
+        Connects the streaming chunks into a single ChatMessage.
+
+        :param chunk: The last chunk returned by the OpenAI API.
+        :param chunks: The list of all chunks returned by the OpenAI API.
+        """
+
+        # if it's a tool call , we need to build the payload dict from all the chunks
+        if bool(chunks[0].meta.get("tool_calls")):
+            tools_len = len(chunks[0].meta.get("tool_calls", []))
+
+            payloads = [{"arguments": "", "name": ""} for _ in range(tools_len)]
+            for chunk_payload in chunks:
+                deltas = chunk_payload.meta.get("tool_calls") or []
+
+                # deltas is a list of ChoiceDeltaToolCall or ChoiceDeltaFunctionCall
+                for i, delta in enumerate(deltas):
+                    payloads[i]["id"] = delta.id or payloads[i].get("id", "")
+                    if delta.function:
+                        payloads[i]["name"] += delta.function.name or ""
+                        payloads[i]["arguments"] += delta.function.arguments or ""
+
+            # handle potential malformed JSON strings
+            tools_calls = [
+                ToolCall(id=payload["id"], tool_name=payload["name"], arguments=json.loads(payload["arguments"]))
+                for payload in payloads
+            ]
+            complete_response = ChatMessage.from_assistant(tool_calls=tools_calls)
+        else:
+            complete_response = ChatMessage.from_assistant("".join([chunk.content for chunk in chunks]))
+        complete_response.meta.update(
+            {
+                "model": chunk.model,
+                "index": 0,
+                "finish_reason": chunk.choices[0].finish_reason,
+                "usage": {},  # we don't have usage data for streaming responses
+            }
+        )
+        return complete_response
+
+    def _build_message(self, completion: ChatCompletion, choice: Choice) -> ChatMessage:
```

Comment:
Let's calls this `convert_chat_completion_to_chat_message`.

Reply:
done